Handle RTL text

chrisgarrity opened the pull request:
### Resolves

_What Github issue does this resolve (please include link)?_
resolves #1675 and #1683 (maybe also the rtl problems in edge, but I haven't checked)

Fixes issues related to dropdown menus (right align instead of left align), and bi-directional blocks (e.g., go to x: %1 y:%2, the ‘x: %1 y:%2’ is LTR, but should not be treated as LTR, when rendered in Hebrew it should be ‘%2 :y %1 :x’)

### Proposed Changes

Sets direction `rtl` for blocklyText when injection is RTL. And redefines the rtl dropdown menuitem styles to look better in RTL

### Reason for Changes

With these changes the dropdown menu and blocks text displays better:
![screen shot 2018-08-24 at 4 59 57 pm](https://user-images.githubusercontent.com/399209/44609152-9af55c80-a7c4-11e8-9125-22674e8054c4.png)
![screen shot 2018-08-24 at 5 01 15 pm](https://user-images.githubusercontent.com/399209/44609158-a052a700-a7c4-11e8-9765-6aa042ce295f.png)

**Without these changes:**
The text is left aligned even though the padding has been reversed:
![screen shot 2018-08-24 at 5 00 37 pm](https://user-images.githubusercontent.com/399209/44609174-b2cce080-a7c4-11e8-9b4c-9c3d39b3b5f7.png)

The way mixed latin and hebrew characters are rendered makes nonsense of the block (note the 'x:' at the start of the block):
![screen shot 2018-08-24 at 5 00 57 pm](https://user-images.githubusercontent.com/399209/44609185-b6606780-a7c4-11e8-957f-40eb5dae63d5.png)

I did try `text-align: right` instead of setting direction for the dropdown, but it still does strange things with mixed bidirectional text:
![screen shot 2018-08-24 at 5 50 40 pm](https://user-images.githubusercontent.com/399209/44609632-7f8b5100-a7c6-11e8-8865-ce3252f13a2f.png)

While the one with direction looks correct:
![screen shot 2018-08-24 at 5 52 31 pm](https://user-images.githubusercontent.com/399209/44609643-8b771300-a7c6-11e8-8279-becc1c7a08f4.png)

/cc @rachel-fenichel while this makes scratch blocks look better, are there horrible side-effects in blockly I'm not aware of?

chrisgarrity committed:
Handle RTL text

Fixes issues related to dropdown menus (right align instead of left align), and bi-directional blocks (e.g., go to x: %1 y:%2, the ‘x: %1 y:%2’ is LTR, but should not be treated as LTR, when rendered in Hebrew it should be ‘%2 :y %1 :x’)

ericrosenbaum reviewed (approved):
LGTM!

chrisgarrity commented:
@AnmAtAnm @rachel-fenichel any comments on these changes for RTL blocklyText?

chrisgarrity commented:
I wondered about splitting out the blocklyDropDownDiv and just overriding the direction on that class, but wasn't sure if we were using the WidgetDiv somewhere that would matter. 

I was just looking at the source in the blockly playground, and somewhere the `blocklyWidgetDiv` got a `style="direction: rtl` added. That would make all the difference!
```
<div class="blocklyWidgetDiv" style="direction: rtl; display: block; height: 69px; top: 127px; left: 295.831px;">
    <div class="goog-menu goog-menu-vertical blocklyDropdownMenu" role="menu" aria-haspopup="true" tabindex="0" style="user-select: none;">
...
    </div>
</div>
```

Also, the text in the blocks has invisible right-to-left markers: `&rlm;` in front of the hebrew text. The question is how did they get there - automatically, or as part of the translation?

chrisgarrity commented:
Yup, widgetDiv has an rtl parameter to set the direction: 
https://github.com/LLK/scratch-blocks/blob/b537fc4b99b42fadf60fe7cf94f3156719c92bf6/core/widgetdiv.js#L110

but dropdownDiv doesn't:
https://github.com/LLK/scratch-blocks/blob/b537fc4b99b42fadf60fe7cf94f3156719c92bf6/core/dropdowndiv.js#L222

I'm not sure where the right-left-marker is coming from though.

chrisgarrity committed:
Set dropdownDiv direction

Set the direction based on the rtl of the block that owns the menu.

Note the css for dropdowns is completely different from blockly (`float` vs `absolute`) as a result of this PR from a couple of years ago: https://github.com/LLK/scratch-blocks/pull/632/

rachel-fenichel commented:
Is this the right-to-left marker you're seeing? https://github.com/LLK/scratch-blocks/blob/develop/core/field.js#L607

chrisgarrity commented:
Yes, that's the marker, and it's definitely getting added to the scratch blocks, but it's not handling mixed hebrew with english. I looked through the blockly demo, and I was only able to find one place where English and Hebrew is mixed in the same blocklyText :
![screen shot 2018-08-29 at 2 09 19 pm](https://user-images.githubusercontent.com/399209/44806717-b336f400-ab95-11e8-9a52-fa1f8bbb2f2e.png)

Everywhere else that I saw they are in separate blocklyText items, for example:
![screen shot 2018-08-29 at 2 13 08 pm](https://user-images.githubusercontent.com/399209/44806759-cf3a9580-ab95-11e8-96d4-e8f687b0cd62.png)

Here's what's in the json for 'go to x, y'
```
"MOTION_GOTOXY": "קפוץ אל x: %1 y: %2",
```
That ends up as two pieces of blockly text (and I can't even copy paste them as examples without things ending up out of order)

Even though a `&rlm` gets added to the string for the 'goto x:', the 'x:' ends up on the right of the hebrew (i.e. at the 'end' in LTR)

I did find a problem with just overriding the direction... negative number inputs appear as `10-`

chrisgarrity commented:
Did you know that when you enter a negative number in blockly in hebrew the minus shows up on the right:
<img width="206" alt="screen shot 2018-08-29 at 6 53 46 pm" src="https://user-images.githubusercontent.com/399209/44820151-ae863600-abbd-11e8-88fc-61e30dc78ca1.png">

but then moves to the left when you stop editing.
<img width="192" alt="screen shot 2018-08-29 at 6 54 00 pm" src="https://user-images.githubusercontent.com/399209/44820155-b1812680-abbd-11e8-819f-393ebd934d28.png">

I do have a new version of scratch-blocks that uses the right-left-embed unicode character instead of the rlm, and everything looks better, except it has the same 'minus sign on the right when editing' as blockly:
<img width="260" alt="screen shot 2018-08-29 at 7 11 35 pm" src="https://user-images.githubusercontent.com/399209/44820646-c9f24080-abbf-11e8-8cab-a492b8b2ebc8.png">

So it's getting closer!

rachel-fenichel commented:
Yes, and i think I filed a bug against blockly for that a while back, but I don't remember where the conversation went.

chrisgarrity commented:
Ah, the new code is actually forcing numbers to be LTR. I can make it put the negative on the right all the time if that's the desired outcome: https://github.com/google/blockly/issues/1802

chrisgarrity committed:
use right-left-embed (rle) unicode to force RTL text

Force RTL for text, but keep numbers LTR in `field.js` instead of overriding direction for all blocklyText. Overriding blocklyText results in negative numbers having the sign on the right instead of the left.

chrisgarrity committed:
Revert changes no longer needed.

chrisgarrity commented on core/field.js:
```diff
@@ -603,8 +603,12 @@ Blockly.Field.prototype.getDisplayText_ = function() {
+    if (this.sourceBlock_.editable_ && this.sourceBlock_.type === 'math_number') {
+      text += '\u200E';
+    } else {
+      text = '\u202B' + text + '\u202C';
```
Can have negative numbers with the '-' on the right (`10-`) by just using line 610.

guyhed replied:
We have a big problem with the blocks!
It seems that {direction:rtl} sould be use only on some blocks. On other blocks it causes things like:
![capture](https://user-images.githubusercontent.com/1233716/44848803-c95db680-ac60-11e8-96de-6fdb720e3673.PNG)

And it also causes stuff like the "10-" instead of "-10".

Note that all math notation and expression should be purely LTR. 
Maybe the solution is not to use RTL at all in the client, and revise the translation/json so that it will be rendered correctly is the client. 
Another option is to mark expression in the translation that should be rendered RTL (maybe by some prefix), and the rest will be LTR  by default.

chrisgarrity replied:
@guyhed in this version most things are better - for example, < and > give the expected results. Subtraction and division can be handled in the translation by switching the order of the arguments in the translation knowing that they will be rendered RTL.

chrisgarrity commented:
@rachel-fenichel I think I'm going to land this PR as it is now. There's still the problem while editing numbers in RTL, but that's an existing issue in blockly that we can deal with separately. This fixes the bidi text and dropdown menus so it's worth getting in.

chrisgarrity committed:
gui needs numbers wrapped as LTR

While just the left-right-marker seems to work in the vertical playground. scratch-gui needs numbers wrapped in the bidi-embed unicode.

rachel-fenichel commented:
Sounds good.

ericrosenbaum reviewed (approved):
LG!
